packages/composite-checkout/src/components/coupon.js

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Internal dependencies
+ */
+import { useLocalize } from '../lib/localize';
+import joinClasses from '../lib/join-classes';
+import Field from './field';
+import Button from './button';
+import Notice from './notice';
+
+export default function Coupon( { id, couponAdded, className, isCouponFieldVisible } ) {
+	const localize = useLocalize();
+	const [ isApplyButtonActive, setIsApplyButtonActive ] = useState( false );
+	const [ couponFieldValue, setCouponFieldValue ] = useState( '' );
+	const [ hasCouponError, setHasCouponError ] = useState( false );
+	const [ isNoticeVisible, setIsNoticeVisible ] = useState( false );
+	const [ isCouponApplied, setIsCouponApplied ] = useState( false );
+
+	useCouponField(
+		id,
+		couponFieldValue,
+		setHasCouponError,
+		couponAdded,
+		setIsCouponApplied,
+		setIsNoticeVisible
+	);
+
+	if ( ! isCouponFieldVisible || isCouponApplied ) {
+		if ( isNoticeVisible ) {
+			return <NoticeComponent />;
+		}
+
+		return null;
+	}
+
+	return (
+		<CouponWrapper className={ joinClasses( [ className, 'coupon' ] ) }>
+			<Field
+				id={ id }
+				placeholder={ localize( 'Enter your coupon code' ) }
+				isError={ hasCouponError }
+				errorMessage={
+					hasCouponError
+						? localize( "We couldn't find your coupon. Please check your code and try again." )
+						: null
+				}
+				onChange={ input => {
+					handleFieldInput( input, setCouponFieldValue, setIsApplyButtonActive, setHasCouponError );
+				} }
+			/>
+
+			{ isApplyButtonActive && (
+				<ApplyButton
+					buttonState="secondary"
+					onClick={ value => {
+						handleApplyButtonClick(
+							value,
+							setHasCouponError,
+							couponAdded,
+							setIsCouponApplied,
+							setIsNoticeVisible
+						);
+					} }
+				>
+					{ localize( 'Apply' ) }
+				</ApplyButton>
+			) }
+
+			{ isNoticeVisible && <NoticeComponent /> }
+		</CouponWrapper>
+	);
+}
+
+Coupon.propTypes = {
+	id: PropTypes.string.isRequired,
+	couponAdded: PropTypes.func,
+};
+
+const animateIn = keyframes`
+  from {
+    opacity: 0;
+    transform: translateX(8px);
+  }
+
+  to {
+    opacity: 1;
+    transform: translateY(0);
+  }
+`;
+
+const CouponWrapper = styled.div`
+	margin: ${props => props.marginTop} 0 0 0;
+	padding-top: 0;
+	position: relative;
+	width: 100%;
+`;
+
+const ApplyButton = styled( Button )`
+	position: absolute;
+	top: 4px;
+	right: 4px;
+	padding: 7px;
+	animation: ${animateIn} 0.2s ease-out;
+`;
+
+function NoticeComponent() {
+	const localize = useLocalize();
+
+	return (
+		<Notice type="success">
+			{ localize( 'Your coupon for $20 has been applied to your cart' ) }
+		</Notice>
+	);
+}
+
+function handleApplyButtonClick(
+	value,
+	setHasCouponError,
+	couponAdded,
+	setIsCouponApplied,
+	setIsNoticeVisible
+) {
+	//TODO: Validate coupon field and replace couponAdded in the following if statement
+	if ( couponAdded ) {
+		setIsNoticeVisible( true );
+		setIsCouponApplied( true );
+		setTimeout( () => {
+			setIsNoticeVisible( false );
```

Comment:
Using `setTimeout` can cause unexpected behavior sometimes since the timeout is based on the event loop and not real time. It also makes things tricky to test. I'm curious what the purpose of this timeout is? If it's for animation purposes, is there a way we can do the timing via CSS instead of JS?

Reply:
It's for removing the component from the DOM after it's displayed for the time it needs. We can technically use CSS to make it disappear but then it'll linger around in the DOM. Any other thoughts of how we might do that?

Reply:
Hm... ok, so let me better understand what's supposed to happen here. You open the coupon field, type something and submit, then the coupon gets added somewhere and also we get a notice displayed somewhere that appears briefly, right?

I wonder if it would be better for now to just not have the notice at all. Here's my reasoning; tell me what you think. Each host page will have its own notification system in place. We can have our own notifications, of course, but it may often be a good idea to use the ones the host provides. But we'll need to wire that in once we have some more real shopping cart behavior and so maybe we should wait until that's ready. Even if we do want to use our own notification for this, it makes sense to me to spend some time architecting that part of the system and it could be done in its own PR.

Reply:
> Hm... ok, so let me better understand what's supposed to happen here. You open the coupon field, type something and submit, then the coupon gets added somewhere and also we get a notice displayed somewhere that appears briefly, right?

yes, that's right. 

> I wonder if it would be better for now to just not have the notice at all. Here's my reasoning; tell me what you think. Each host page will have its own notification system in place. We can have our own notifications, of course, but it may often be a good idea to use the ones the host provides. But we'll need to wire that in once we have some more real shopping cart behavior and so maybe we should wait until that's ready. Even if we do want to use our own notification for this, it makes sense to me to spend some time architecting that part of the system and it could be done in its own PR.

Yeah — that makes a lot of sense to me. I'll take the notification out.